### ISSUE TITLE
Atari: calling cgetc() resets Bitmap mode DLIST register $0942 to 0

### ISSUE THREAD

**Issue by 8bit-Dude** [cc65/cc65#1324]
This one is for Chris Groessler I think...

By default, cc65 generate a Text DLIST in page 9 (starting @ $0920). 

When switching to bitmap mode, I use the same start address for my DLIST (writing over a portion of the now unused text memory).

However, calling cgetc() seems to always reset $0942 to 0 (I think it may have something to do with the cursor...).

Would it be possible to move whather $0942 is used for to an address outside of the Bitmap mode DLIST (which goes up-to $09f1)?

**Comment by 8bit-Dude** [comment]
Note: I tried cursor(0) before calling cgetc(), but I get the same result.... (some register set to 0 in my DLIST)

**Comment by 8bit-Dude** [comment]
As a temporary solution, I have set OLDADR ($5E) to an address above my DLIST.
The issue is that setcursor will reset this char, even if cursor() is set to 0.

**Comment by groessler** [comment]
> By default, cc65 generate a Text DLIST in page 9 (starting @ $0920).

This is not true in general. On the `atari` target the dlist is at top-of-ram (set-up by the ROM), and on the `atarixl` target the location depends on the start address of the program. 
So you seem to be using the `atarixl` program. (I knew that before from our PMs, I'm writing this here just for others to know...)

I guess you are not using a text mode and are calling `cgetc()`.
Can you implement the following change to `libsrc/atari/setcursor.s`, recompile the cc65 lib, and check out if that helps?

```
@@ -37,11 +37,11 @@
 cont:   stx     CRSINH          ; update system variable
 
         beq     turnon
-        and     #$7f            ; clear high bit / inverse flag
+        ;and     #$7f            ; clear high bit / inverse flag
 finish: sta     (OLDADR),y      ; update on-screen display
         rts
 
-turnon: ora     #$80            ; set high bit / inverse flag
-        bne     finish
+turnon: ;ora     #$80            ; set high bit / inverse flag
+        jmp     finish
 
 .endproc
```

**Comment by 8bit-Dude** [comment]
> This is not true in general. On the `atari` target the dlist is at top-of-ram (set-up by the ROM), and on the `atarixl` target the location depends on the start address of the program.
> So you seem to be using the `atarixl` program. (I knew that before from our PMs, I'm writing this here just for others to know...)

Ah yes, that's correct! (what I really meant by "default" was the mechanism by which atarixl target locates the DLIST). In actual fact, if one uses the atarixl config address to locate a BMP DLIST, then the same problem with cursor will always occur.

I will try your changes and see what happens, thanks for giving it a look.

P.S: Regarding the use of cgetc(), is there another function that retrieves keyboard entries in non-text mode?

**Comment by groessler** [comment]
> Regarding the use of cgetc(), is there another function that retrieves keyboard entries in non-text mode?

I don't know of one. But this doesn't mean much. One would have to analyze the ROM code looking for something. And, unofficial entry points should/must be avoided.

**Comment by oliverschmidt** [comment]
I'm classifying this issue as a bug as it seems a reasonable expectation from my POV to be able to use cgetc() independently from the current video mode - especially when cursor(0) was called before. Please let me know if I'm missing the point...

**Comment by dmsc** [comment]
Hi!

> I'm classifying this issue as a bug as it seems a reasonable expectation from my POV to be able to use cgetc() independently from the current video mode - especially when cursor(0) was called before. Please let me know if I'm missing the point...

The problem here is that the user did not change the graphics mode via the OS, so we can't know that we are not supposed to update the cursor.

Also, calling "cursor()" does not alter the current cursor on screen, it simply sets a flag so that next input/output to the screen actually draws or clears the cursor.

**Comment by 8bit-Dude** [comment]
> The problem here is that the user did not change the graphics mode via the OS, so we can't know that we are not supposed to update the cursor.
> 
> Also, calling "cursor()" does not alter the current cursor on screen, it simply sets a flag so that next input/output to the screen actually draws or clears the cursor.

(1) Some widely used graphic modes are not supported by the OS (specifically, adding an extra line at the bottom of screen). But I fail to see how setting up the DLIST with OS would change things. There should not be any screen buffer when the system is in bitmap mode.

(2) cgetc() is calling setcursor() and setting the underlying byte to 0 regardless of the cursor() state. 

I will test Chris's suggested changes, and report back.

**Comment by dmsc** [comment]
Hi!

> 
> (1) Some widely used graphic modes are not supported by the OS (specifically, adding an extra line at the bottom of screen). But I fail to see how setting up the DLIST with OS would change things. There should not be any screen buffer when the system is in bitmap mode.

Problem is, you put your display list over the old screen memory area, and did not tell the OS that this is not the screen area anymore.

> 
> (2) cgetc() is calling setcursor() and setting the underlying byte to 0 regardless of the cursor() state.
> 

Yes, because "cgetc()" does not know the old state of the cursor, so it needs to clear the cursor before reading the key. And to clear the cursor, it writes to the cursor memory location the old byte.

The solution is simple: If you set a custom graphics mode, you need to set SAVMSC and OLDADR to the top of your screen memory and OLDCHR to the value under that location.

> I will test Chris's suggested changes, and report back.

That wont work, as the old cursor will still be cleared.

Have Fun!

**Comment by oliverschmidt** [comment]
Now I'm even more puzzled. My understanding is that there's in general no visible cursor in CONIO. Only while waiting for user input a cursor is - optionally - drawn. So with cursor(0) there's no reason for cgetc() to fiddle with the screen in any way.

**Comment by 8bit-Dude** [comment]
> The solution is simple: If you set a custom graphics mode, you need to set SAVMSC and OLDADR to the top of your screen memory and OLDCHR to the value under that location.

My bitmap DLIST runs from $0920 to $09f1. 
So do you mean that I should set SAVMSC  and OLDADR  to $09f2 and OLDCHR to $09f3?

**Comment by dmsc** [comment]
Hi!

> Now I'm even more puzzled. My understanding is that there's in general no visible cursor in CONIO. Only while waiting for user input a cursor is - optionally - drawn. So with cursor(0) there's no reason for cgetc() to fiddle with the screen in any way.

Then, the bug is in an unrelated place.

Currently CONIO in the Atari *always* displays a cursor after any character output, it moves the cursor on "gotoxy" and clears it on any output *after* using "cursor(0)". It basically mimics the OS cursor usage.

Reading the docs, it seems that the intended behavior is that the cursor is only displayed on input. So, the current CONIO implementation in the Atari could be vastly simplified and made faster.

@groessler , what do you think?

**Comment by dmsc** [comment]
Hi!

> My bitmap DLIST runs from $0920 to $09f1.
> So do you mean that I should set SAVMSC and OLDADR to $09f2 and OLDCHR to $09f3?

If you don't plan on using character output, it is probably easier ti set SAVMSC, OLDADR to 0 and OLDCHR (a byte) to 0. That would put a 0 into address 0, but address 0 is not normally used.

Have Fun!

**Comment by oliverschmidt** [comment]
> Reading the docs, it seems that the intended behavior is that the cursor is only displayed on input.

Indeed!

> @groessler , what do you think?

Me too!

**Comment by groessler** [comment]
This cursor "dance" is there, since the ROM code on the Atari always displays an "on" cursor.
But with cc65 programs "we" (in fact, "I") only want a cursor visible on keyboard input (to be in line with c64 and other targets).

If someone has a better idea to implement that - or people want or don't mind an always-visible cursor - it can be changed.

Please note, that for input (`getc`) the ROM routines are used, but on output the ROM routines or "direct access to screen memory" could be used (stdio vs. `cputc`).

**Comment by polluks** [comment]
Explaining some motivation
* stdio should use the ROM/OS for features (e.g. XEP80)
* conio should be fast and uniform

**Comment by oliverschmidt** [comment]
@groessler: Thanks for the background info :-)

So from my POV the "unusual" thing then is that cgetc() uses ROM code. Usually STDIO uses ROM code for both input and output while CONIO doesn't use ROM code for neither input nor output.

I presume there are compelling reasons for using ROM code for CONIO input (aka cgetc()/kbhit()).

**Comment by groessler** [comment]
@oliverschmidt: The Commodores, I've checked c64 and vic20, also use ROM code in `cgetc`.

Regarding the cursor handling, I'm seeing similar things in the CBMs and AppleII.

**Comment by groessler** [comment]
@8bit-Dude, did you try with the change I've posted?

**Comment by oliverschmidt** [comment]
> The Commodores, I've checked c64 and vic20, also use ROM code in cgetc.

Thanks for pointing out.

> Regarding the cursor handling, I'm seeing similar things in the CBMs and AppleII.

I proved once more above that I have no clue regarding the CBMs but I have no idea what you are referring to regarding the Apple II. Please elaborate!

**Comment by groessler** [comment]
See the `jsr putchardirect` in `libsrc/apple2/cgetc.s`, which I think update screen memory.

**Comment by oliverschmidt** [comment]
But those JSRs are skipped when cursor is off!

**Comment by groessler** [comment]
I see now.

Looks like the Atari version is broken in this regard, that it unconditionally touches screen memory. I need to investigate.

But, @8bit-Dude, it's still interesting if the workaround works for you.

**Comment by dmsc** [comment]
Hi!

IMHO, best option would be to simply remove all cursor support on cputc and all other conio output functions. This would make the code smaller and faster, and conform to the documentation.

And to clarify things: on the Atari, the OS get character does not writes or erases a cursor, so using the OS for cgetc is always OK. The reason for using the OS is that is smaller, and handles all different Atari 8-bit computers.

Have Fun!

**Comment by groessler** [comment]
@dmsc: I think "remove all cursor support on cputc and all other conio output functions" is a no-go. Don't you expect a cursor when the computer waits for input?
And with cursor support, conio and stdio can me more or less mixed within limits (I know that is not officially supported, but it's convenient).

**Comment by dmsc** [comment]
Hi!

> @dmsc: I think "remove all cursor support on cputc and all other conio output functions" is a no-go. Don't you expect a cursor when the computer waits for input?

Yes, and that is the key: You should draw the cursor only when waiting for input (on cgetc) if requested, just oposite of what it is done currently, that the cursor is drawn on each put call and erased on cgetc if the user did not ask for a cursor.

> And with cursor support, conio and stdio can me more or less mixed within limits (I know that is not officially supported, but it's convenient).

And also does not always work - if you install any extended E: handler (for example, an 80 column driver, or an I/O redirection to a file) then you can't mix both.

So, IMHO, it would be ok to remove that capability, in the interest of making applications smaller and faster.

Have Fun!

**Comment by oliverschmidt** [comment]
> I think "remove all cursor support on cputc and all other conio output functions" is a no-go. Don't you expect a cursor when the computer waits for input?

@groessler: I imagine that there's a misunderstanding: @dmsc explicitly refered to the conio **_output_** functions. I'm sure that nobody questions the usefulness of a cursor during input.

**Comment by oliverschmidt** [comment]
> You should draw the cursor only when waiting for input (on cgetc) if requested [...]

That's at least my expectation for a CONIO implementation.

> And with cursor support, conio and stdio can me more or less mixed within limits [...]

My POV:

STDIO is about maximum compliance with the usual behavior on a given target.
CONIO is about high/maximum performance **_and_** about maximum consistency across targets.

If I'd have to choose between interoperability of CONIO with STDIO and consistence of CONIO with expectations and other targets then I'd choose the latter.

Apart from that I imagine that are ways to make the interoperability of CONIO with STDIO regarding cursor handling "good enough". In case I got it right that the cursor is usually on, I'could see e.g. this:

* A CONIO constructor turns the cursor off
* A CONIO destructor turns the cursor on
* A function _cursor() (note the underscore) turns the cursor on/off

So if one wants to use STDIO in a program using CONIO one has to wrap (groups of) STDIO calls _cursor(1) / _cursor(0).

**Comment by dmsc** [comment]
Hi!

> Apart from that I imagine that are ways to make the interoperability of CONIO with STDIO regarding cursor handling "good enough". In case I got it right that the cursor is usually on, I'could see e.g. this:
> 
> * A CONIO constructor turns the cursor off
> * A CONIO destructor turns the cursor on

That would work, but see details bellow.

> * A function _cursor() (note the underscore) turns the cursor on/off

The Atari OS assumes that the cursor is always on - it uses three ZP locations to store the cursor address (OLDADR) and the memory value "under" the cursor (OLDCHR). On each character output, it overwrites that address with the value, in C: ` *OLDADR = OLDCHR `. So, you can't "turn the cursor off" in the OS without writing something to the screen.

> 
> So if one wants to use STDIO in a program using CONIO one has to wrap (groups of) STDIO calls _cursor(1) / _cursor(0).

No, that would not work. It should be another pair of calls, like "stop_conio()" and "start_conio()".

Basically, in C:
```c
#include <atari.h>
void start_conio(void) {
    // Force disabling the OS cursor.
    * OS.oldadr = OS.oldchr;
}
void stop_conio(void) {
    // Update OS cursor locations
    OS.oldadr = OS.savmsc + OS.rowcrs * 40 + OS.colcrs;
    OS.oldchr = * OS.oldadr;
}
```

Have Fun!

**Comment by oliverschmidt** [comment]
Maybe yet again a misunderstanding(?) The `_cursor()` I mention above has nothing to go with the CONIO `cursor()`. It seems to me that my `_cursor(1)` equals your `stop_conio()` and my `_cursor(0)` equals your `start_conio()`. At least the comments in your code suggest that perspective...

**Comment by dmsc** [comment]
Hi!

> Maybe yet again a misunderstanding(?) The `_cursor()` I mention above has nothing to go with the CONIO `cursor()`. It seems to me that my `_cursor(1)` equals your `stop_conio()` and my `_cursor(0)` equals your `start_conio()`. At least the comments in your code suggest that perspective...

Yes, I see, I was confused with the CONIO "cursor()" function. And yes, it is the same - except that "stop_conio()" does not draws the cursor - that is left for the next call to OS output functions.

I prefer the "start/stop conio" names because the user could be confused with the "_cursor()" names, thinking that the functions are only necessary if the cursor is visible. Bus as I said before, the OS *always* clears the cursor on output, even if it is not drawn.

Have Fun!

**Comment by oliverschmidt** [comment]
Thanks for the clarification. Anyhow, this was just supposed to be a rough sketch illustrating that at least from my POV something awkward like having to call additional proprietary functions for switching between CONIO and STDIO would be preferable over having CONIO behave the way it does now.

**Comment by groessler** [comment]
Since there were since almost 20 yrs no complaints/problems with the current implementation (*) I tend to leave it as it is. Unless @8bit-Dude reports back that the workaround I had posted doesn't work for him.

@8bit-Dude ?

* @Fabrizio-Caruso had a problem when using mode 1 (in BASIC talk) and it worked with the changed `setcursor.s`.

**Comment by oliverschmidt** [comment]
> Since there were since almost 20 yrs no complaints/problems with the current implementation (*) [...]

Sorry, but I think this isn't true. I'm pretty sure that I complained about it several times, e.g. in the context of the Contiki GUI.

Apart from that - even if there were no complaints in the past: If now everybody would consent that a different approach would be preferable than that's enough to make the different approach desirable. Or how are new things expected to come into the world otherwise? ;-)

If you are both motived enough and have the time budget to spent for implementing the different approach is of course a totally different question.

**Comment by 8bit-Dude** [comment]
> @8bit-Dude ?
> * @Fabrizio-Caruso had a problem when using mode 1 (in BASIC talk) and it worked with the changed `setcursor.s`.

Sorry for a very late reply on this Chris. I tried your proposed change but it did not fix the issue. I still get a corrupted DLINE:

I could however get the issue fixed without modifying the Atari lib, with dmsc's suggestion:

	POKEW(SAVMSC, 0);
	POKEW(OLDADR, 0);
	POKE(OLDCHR, 0);[](url)

**Comment by groessler** [comment]
I'm going to look into this.
